### PR TITLE
DCS-1348 Tidying up state handling code and adding ability to store SearchDetails

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: integration_tests/integration/choosePrisoner.spec.ts [hunk 1]
```diff
@@ -175,13 +175,15 @@ context('Choose Prisoner', () => {
...
     rowNumber: number
   ) {
-    cy.task('stubExpectedArrival', stub)
+    cy.task('stubExpectedArrival', arrival)
     choosePrisonerPage.arrivalFrom(arrivalType)(rowNumber).confirm().should('exist').click()
-    Page.verifyOnPage(ConfirmArrivalPage).name().should('contain.text', `${stub.firstName} ${stub.lastName}`)
+    Page.verifyOnPage(ConfirmArrivalPage)
+      .existingName()
+      .should('contain.text', `${arrival.firstName} ${arrival.lastName}`)
     cy.go('back')
```
Comment (andrewrlee) [inline integration_tests/integration/choosePrisoner.spec.ts:187]: Fixing this as existing code didn't type check

File: package.json [hunk 2]
```diff
@@ -19,7 +19,7 @@
     "watch-node-feature": "export $(cat feature.env) && nodemon --watch dist/ $NODE_DEBUG_OPTION dist/server.js | bunyan -o short",
     "start-feature:dev": "npm run build && concurrently -k -p \"[{name}]\" -n \"Views,TypeScript,Node,Sass\" -c \"yellow.bold,cyan.bold,green.bold,blue.bold\" \"npm run watch-views\" \"npm run watch-ts\" \"npm run watch-node-feature\" \"npm run watch-sass\"",
     "record-build-info": "node ./bin/record-build-info",
-    "lint": "eslint . --cache --max-warnings 0",
+    "lint": "eslint . --cache --max-warnings 0 && npm run typecheck",
```
Comment (andrewrlee) [inline package.json:22]: ensuring type check happens as part of linting -this is run by the build so should prevent code that doesn't compile being committed

File: server/routes/bookedtoday/arrivals/searchforexisting/state.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+const SearchDetailsCodec: Codec<SearchDetails> = {
+  write: (value: SearchDetails): Record<string, string> => ({ ...value }),
+
+  read(record: Record<string, unknown>): SearchDetails {
+    assertHasStringValues(record, ['firstName', 'lastName', 'movementReasonCode'])
+    assertHasOptionalStringValues(record, ['dateOfBirth', 'prisonNumber', 'pncNumber'])
+
+    return {
+      firstName: record.firstName,
+      lastName: record.firstName,
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/searchforexisting/state.ts:21]: this is wrong and is fixed in follow up

File: server/routes/bookedtoday/arrivals/state.ts [hunk 4]
```diff
@@ -52,16 +35,7 @@ export const SexCodec: Codec<string> = {
...
-
-export const setSex = (res: Response, data: string): void => setState(SEX_COOKIE_NAME, SexCodec)(res, data)
-
-export const getSex = (req: Request): string | undefined => getState(SEX_COOKIE_NAME, SexCodec)(req)
-
-export const ensureSexPresentMiddleware =
-  (redirectUrl: string): RequestHandler =>
-  (req: Request, res: Response, next: NextFunction) => {
-    return isStatePresent(SEX_COOKIE_NAME)(req) ? next() : res.redirect(redirectUrl)
-  }
```
Comment (andrewrlee) [inline server/routes/bookedtoday/arrivals/state.ts:67]: extracted this into a generic `stateOperations`  to cut down boiler plate

File: server/utils/utils.ts [hunk 5]
```diff
@@ -48,3 +48,18 @@ export function assertHasStringValues<K extends keyof Record<string, unknown>>(
     throw Error(`Missing or invalid keys: ${invalidKeys}`)
   }
 }
+
+export function assertHasOptionalStringValues<K extends string>(
+  obj: unknown,
+  keysToCheck: K[]
+): asserts obj is Record<K, string | undefined> {
```
Comment (andrewrlee) [inline server/utils/utils.ts:55]: version of `assertHasStringValues` which copes with fields being absent. 

this is a type assertion to allow safe access of fields on unknown types:
```
const a: unknown = ...
assertHasOptionalStringValues(a, ['name'])
console.log(a.name)   // usually you wouldn't be able to access this
```

**Review by GurnankCheema — APPROVED** [review]
LGTM

**Pull request merged**